drivers/interrupt_controller/intc_miwu.c

```diff
@@ -0,0 +1,368 @@
+/*
+ * Copyright (c) 2020 Nuvoton Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nuvoton_npcx_miwu
+
+/**
+ * @file
+ * @brief Nuvoton NPCX MIWU driver
+ *
+ * The device Multi-Input Wake-Up Unit (MIWU) supports the Nuvoton embedded
+ * controller (EC) to exit 'Sleep' or 'Deep Sleep' power state which allows chip
+ * has better power consumption. Also, it provides signal conditioning such as
+ * 'Level' and 'Edge' trigger type and grouping of external interrupt sources
+ * of NVIC. The NPCX series has three identical MIWU modules: MIWU0, MIWU1,
+ * MIWU2. Together, they support a total of over 140 internal and/or external
+ * wake-up input (WUI) sources.
+ *
+ * This driver uses device tree files to present the relationship bewteen
+ * MIWU and the other devices in different npcx series. For npcx7 series,
+ * it include:
+ *  1. npcxn-miwus-wui-map.dtsi: it presents relationship between wake-up inputs
+ *     (WUI) and its source device such as gpio, timer, eSPI VWs and so on.
+ *  2. npcx7-miwus-int-map.dtsi: it presents relationship between MIWU group
+ *     and NVIC interrupt in npcx7. Please notice it isn't 1-to-1 mapping.
+ *     For example, here is the mapping between miwu0's group a & d and IRQ7:
+ *
+ *     map_miwu0_groups: {
+ *         parent = <&miwu0>;
+ *         group_ad0: group_ad0_map {
+ *             irq        = <7>;
+ *             group_mask = <0x09>;
+ *         };
+ *         ...
+ *     };
+ *
+ *     It will connect IRQ 7 and intc_miwu_isr0() with the argument, group_mask,
+ *     by IRQ_CONNECT() during driver initialization function. With group_mask,
+ *     0x09, the driver checks the pending bits of group a and group d in ISR.
+ *     Then it will execute related callback functions if they have been
+ *     registered properly.
+ *
+ * INCLUDE FILES: soc_miwu.h
+ *
+ */
+
+#include <assert.h>
+#include <device.h>
+#include <kernel.h>
+#include <soc.h>
+#include <irq_nextlevel.h>
+#include <drivers/gpio.h>
+
+#include "soc_miwu.h"
+#include "soc_gpio.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(intc_miwu, LOG_LEVEL_ERR);
+
+/* MIWU module instances forward declaration */
+static struct device *miwu_devs[];
+
+/* Driver config */
+struct intc_miwu_config {
+	/* miwu controller base address */
+	uint32_t base;
```

Comment:
Similar to other comments, this uint32_t type would be better as a uintprt_t or a specialized typedef

Reply:
Done.